Give more detailed validation errors in VA

jsha opened the pull request:
In particular, differentiate timeouts during connect (which are usually a firewall problem) from timeouts after connect (which are usually a software problem). In the process, refactor the tests and add testing for specific problem detail messages.

This also switches over the HTTP challenge's dialer to use DialContext, and to shave a little bit of headroom off of the context deadline, so that the dial can report its timeout before the overall context expires, which would lead to an overly generic "deadline exceeded" error, which would then get translated (incorrectly) into a "timeout after connect."

There is an additional error case, `Timeout during %s (your server may be slow or overloaded)`, (where `%s` can be `read` or `write`) which doesn't have any unittests. I believe it may not be possible to trigger this, since `read` and `write` timeouts get subsumed by the HTTP or TLS library, but it's worth having as a fallback case. We'll see if it shows up in the logs.

Among the test refactorings, I shortened the timeout on the TLS timeout test to 50ms. Previously this was the long pole making the whole test take 10s. Now it takes ~500 ms overall.

I recommend starting review at https://github.com/letsencrypt/boulder/compare/detailed-va-errors?expand=1#diff-4c51d1d7ca3ec3022d14b42809af0d7eR671 (the changes to `detailedError`), then reviewing the `Dial` -> `DialContext` changes, then the tests.

cpu commented on va/va.go:
```diff
@@ -258,7 +271,8 @@ func (d *http01Dialer) Dial(_, _ string) (net.Conn, error) {
 	d.record.AddressUsed = v4[0]
 	realDialer = d.realDialer()
-	return realDialer.Dial("tcp", address)
+	conn, err := realDialer.DialContext(ctx, "tcp", address)
```
nit: was there a reason to capture the two returns into `conn, err` and then `return` them explicitly vs changing the existing `return realDialer.Dial("tcp", address)` into `return realDialer.DialContext(ctx, "tcp", address)`? I think the latter is a little cleaner (smaller diff anyway :-))

jsha commented:
It looks like this PR introduced a bit of test flakiness; please hold off reviewing while I investigate and fix.

jsha commented:
I believe I've fixed the flakiness and this is ready for review. Two changes:

 - Made sure HTTPPort, HTTPSPort, and TLSPort all had default values. Since these weren't configured in all test cases, some tests would try to connect to `:0`, which seemed to lead to inconsistent results locally vs on Travis.
 - Restore singleDialTimeout after the test that modifies it. I think this was causing other tests to timeout when they shouldn't.

jsha commented:
Oh and I forgot to mention the main flakiness culprit, and the actual hardest to track down:

 - These tests use an unrouteable IP to trigger a dial timeout. Some of the time, that immediately returns "network unreachable" instead of timing out. I don't know why this behavior happens, but it's rare-ish (1 in 20 or so), so I added retries to the tests for when that happens (and also added a detailed error for it).

rolandshoemaker commented on va/va.go:
```diff
@@ -649,13 +672,20 @@ func detailedError(err error) *probs.ProblemDetails {
 			return probs.ConnectionFailure("Connection refused")
+		} else if syscallErr, ok := netErr.Err.(*os.SyscallError); ok &&
+			syscallErr.Err == syscall.ENETUNREACH {
+			return probs.ConnectionFailure("Network unreachable")
```
(Left this comment on #3648 but realized this is actually where it came from)

This seems like the wrong problem type, a network unreachable error is almost definitely indicating something wrong with our network stack and should trigger 500s we want to alert on.
